infrastructure/src/main/java/org/corfudb/infrastructure/management/CompleteGraphAdvisor.java

```diff
@@ -0,0 +1,134 @@
+package org.corfudb.infrastructure.management;
+
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.infrastructure.management.failuredetector.ClusterGraph;
+import org.corfudb.protocols.wireprotocol.ClusterState;
+import org.corfudb.protocols.wireprotocol.failuredetector.NodeRank;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * This class is an implementation of {@link ClusterType} where the ideal state
+ * of the Corfu cluster is a fully connected network (i.e. complete graph) in which there is an
+ * active link amongst all nodes in the cluster. Failed and Healed nodes are recommended so that
+ * the cluster remains fully connected.
+ * <p>
+ * Created by Sam Behnam on 10/27/18.
+ */
+@Slf4j
+public class CompleteGraphAdvisor implements ClusterAdvisor {
+
+    private static final ClusterType CLUSTER_TYPE = ClusterType.COMPLETE_GRAPH;
+
+    @Override
+    public ClusterType getType() {
+        return CLUSTER_TYPE;
+    }
+
+    /**
+     * Provides list of servers from a given layout(epoch) that this implementation of
+     * COMPLETE_GRAPH algorithm has determined as failed. The implementation of the
+     * algorithm in this method is an approach by executing the following steps:
+     * <p>
+     * The failed node is the recommendation of this strategy which their removal
+     * from cluster will lead to a fully connected cluster.
+     * <p>
+     * The following represents the underlying implementation of one algorithm to achieve the
+     * above goal however the clients of this strategy must only rely on the guarantee that removal
+     * of the returned failed node is a recommendation for arriving at a fully connected
+     * cluster and must not rely on the implementation details of the algorithm.
+     * <p>
+     *
+     * @param clusterState        represents the state of connectivity amongst the Corfu cluster
+     *                            nodes from a node's perspective.
+     * @param unresponsiveServers list of unresponsive servers.
+     * @return a server considered as failed according to the underlying strategy.
+     */
+    @Override
+    public Optional<NodeRank> failedServer(
+            ClusterState clusterState, List<String> unresponsiveServers, String localEndpoint) {
+
+        log.trace("Detecting failed nodes for: ClusterState= {} unresponsive servers= {}",
+                clusterState, unresponsiveServers
+        );
+
+        ClusterGraph graph = ClusterGraph.transform(clusterState);
+
+        ClusterGraph symmetric = graph.toSymmetric();
+        Optional<NodeRank> maybeDecisionMaker = symmetric.getDecisionMaker();
```

Comment:
The decision maker always exists. There is always at least the local node, it always has at least one successful connection to itself.
We also have two additional checks to prevent all possible incorrect ways. Decision maker not found if:
 - ClusterGraph is empty, which is an invalid state.
 - the decision maker doesn't have connections, which is also impossible.

Reply:
ok please document that. It will be hard for another reviewer down the line to follow otherwise. Also please mention, decision maker is just an optimization. We could still have multiple decision makers and this should be ok. In a partitioned scenario, a minority side can choose to have a decision maker just that it's decisions will not be used as it does not have consensus.

Reply:
done

Reply:
Can you add comments for this?